File: app/src/main/java/org/sorz/lab/tinykeepass/keepass/KeePassHelper.java
package org.sorz.lab.tinykeepass.keepass;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import org.sorz.lab.tinykeepass.FetchDatabaseTask;

import java.io.File;
import java.util.UUID;
import java.util.stream.Stream;

import de.slackspace.openkeepass.domain.Entry;
import de.slackspace.openkeepass.domain.KeePassFile;

/**
 * Common helper methods related to KeePass.
 */
public class KeePassHelper {
    /**
     * Get all entries without ones in recycle bin.
     * @param keePass opened Keepass database file
     * @return stream of all entries not in recycle bin
     */
    static public Stream<Entry> allEntriesNotInRecycleBin(KeePassFile keePass) {
        if (keePass.getMeta().getRecycleBinEnabled()) {
            UUID recycleBinUuid = keePass.getMeta().getRecycleBinUuid();
            return keePass.getGroups().stream()
                    .filter(group -> !group.getUuid().equals(recycleBinUuid))
                    .flatMap(group -> group.getEntries().stream());
        } else {
            return keePass.getEntries().stream();
        }
    }

    /**
     * @param str string or null
     * @return true if str is not null and not empty
     */
    static public boolean notEmpty(String str) {
        return str != null && !str.isEmpty();
    }

    static public Bitmap getIcon(Entry entry) {
        return BitmapFactory.decodeByteArray(entry.getIconData(), 0,
                entry.getIconData().length);
    }

    static public File getDatabaseFile(Context context) {
        return new File(context.getNoBackupFilesDir(), FetchDatabaseTask.DB_FILENAME);
    }

    static public boolean hasDatabaseConfigured(Context context) {
        return KeePassStorage.get(context) != null || getDatabaseFile(context).canRead();
    }
}

File: app/src/main/java/org/sorz/lab/tinykeepass/keepass/KeePassHelper.kt
package org.sorz.lab.tinykeepass.keepass

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Context.CLIPBOARD_SERVICE
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable


import org.sorz.lab.tinykeepass.FetchDatabaseTask

import java.io.File

import com.kunzisoft.keepass.database.element.Database
import com.kunzisoft.keepass.database.element.Entry
import org.jetbrains.anko.startService
import org.sorz.lab.tinykeepass.PasswordCopingService
import org.sorz.lab.tinykeepass.R
import java.util.stream.Stream
import kotlin.streams.asStream

/**
 * Stream of all entries without ones in recycle bin.
 */
val Database.allEntriesNotInRecycleBin: Sequence<Entry>
    get() = sequence {
        val root = rootGroup ?: return@sequence
        val recycleBinUuid = if (isRecycleBinEnabled) recycleBin?.nodeId else null
        val groups = mutableListOf(root)
        while (groups.isNotEmpty()) {
            val group = groups.removeLast()
            if (group.nodeId != null && group.nodeId == recycleBinUuid)
                continue
            groups.addAll(group.getChildGroups())
            yieldAll(group.getChildEntries())
        }
    }

val Database.allEntriesNotInRecycleBinStream: Stream<Entry>
    get() =
    allEntriesNotInRecycleBin.asStream()


val Context.databaseFile: File get() =
    File(noBackupFilesDir, FetchDatabaseTask.DB_FILENAME)

val Context.hasDatabaseConfigured: Boolean get() =
    KeePassStorage.get(this) != null || databaseFile.canRead()

private val Entry.cleanUrl get() = url.replace("^https?://(www\\.)?".toRegex(), "")

val Entry.urlHostname get() = cleanUrl.split("/".toRegex(), 2).first()
val Entry.urlPath get() = cleanUrl.split("/".toRegex(), 2).getOrNull(1)
        ?.takeIf { it.isNotBlank() }?.let { "/$it" }

fun Entry.copyUrl(context: Context) {
    val clipboardManager = context.getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
    val clipData = ClipData.newPlainText("URL", url)
    clipboardManager.setPrimaryClip(clipData)
}

fun Entry.copyUsername(context: Context) {
    val clipboardManager = context.getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
    val clipData = ClipData.newPlainText(context.getString(R.string.username), username)
    clipboardManager.setPrimaryClip(clipData)
}

fun Entry.copyPassword(context: Context) {
    Intent(context, PasswordCopingService::class.java).apply {
        action = PasswordCopingService.ACTION_COPY_PASSWORD
        putExtra(PasswordCopingService.EXTRA_PASSWORD, password)
        context.startService(this)
    }
}

fun Entry.copyPasswordPostponed(context: Context) {
    Intent(context, PasswordCopingService::class.java).apply {
        action = PasswordCopingService.ACTION_NEW_NOTIFICATION
        putExtra(PasswordCopingService.EXTRA_PASSWORD, password)
        putExtra(PasswordCopingService.EXTRA_USERNAME, username)
        putExtra(PasswordCopingService.EXTRA_ENTRY_TITLE, title)
        context.startService(this)
    }
}
